### PR TITLE
Support multiple frustums

### PR DESCRIPTION
**Pull request by nithinp7**
Collects and sends all editor/play cameras to the 3D Tiles refinement engine.

Builds upon the changes here: https://github.com/CesiumGS/cesium-native/pull/283

Fixes: #224 

MultiFrustums  DebugGame  - Unreal Editor 2021-07-13 17-24-36 (1)

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @nithinp7!

* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] [Cesium for Unreal Samples](https://github.com/CesiumGS/cesium-unreal-samples) works.

**Comment by CesiumBen** [comment]
@nithinp7 Awesome work and demo!

 I've tested these updates with VR and motion and am still getting some tiles unloading, but strangely enough only the ones which are some distance from the origin. The gif below displays it as best as possible, the 1st half of the gif I'm far from the starting position and disappearing tiles occurs more frequently. While I return to the origin point (When I'm grabbing with the hands) the culling is much better and the tiles remain as they should.

Any configuration I should be aware of when testing?

clipping

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 1]
```diff
@@ -782,102 +783,187 @@ void ACesium3DTileset::MarkTilesetDirty() {
...
+      rotation = FRotator(0.0f, 0.0f, 0.0f);
+      break;
+    case LVT_OrthoNegativeYZ:
+      rotation = FRotator(0.0f, 180.0f, 0.0f);
+      break;
+    case LVT_OrthoFreelook:
+    case LVT_Perspective:
+    case LVT_MAX:
+    case LVT_None:
+      rotation = pEditorViewportClient->GetViewRotation();
```
Comment (javagl) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:942]: Is there a reason why these cases are handled individually? 

E.g. when the viewport type is `LVT_OrthoXZ`, I'd expect `GetViewRotation` to exactly return `FRotator(0.0f, -90.0f, 0.0f)`. Or will `GetViewRotation` return different values than the hard-codes ones?

File: Source/CesiumRuntime/Public/Cesium3DTileset.h [hunk 2]
```diff
@@ -521,21 +522,23 @@ class CESIUMRUNTIME_API ACesium3DTileset : public AActor,
...
   void MarkTilesetDirty();
-  Cesium3DTiles::ViewState CreateViewStateFromViewParameters(
-      const FVector2D& viewportSize,
-      const FVector& location,
-      const FRotator& rotation,
-      double fieldOfViewDegrees) const;
 
   struct UnrealCameraParameters {
+    bool isPerspective;
+    bool isOrthographic;
```
Comment (javagl) [inline Source/CesiumRuntime/Public/Cesium3DTileset.h:528]: Can `isPerspective` and `isOrthographic` ever have the same value? The frustum type is derived from that with 
```
  if (camera.isOrthographic) {
    frustumType = Cesium3DTiles::ViewState::FrustumType::Orthographic;
  } else {
    frustumType = Cesium3DTiles::ViewState::FrustumType::Perspective;
  }
```
and the method where they are obtained is implemented as
```
bool FEditorViewportClient::IsOrtho() const
{
	return !IsPerspective();
}
```
so I think these fields might be redundant...

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 3]
```diff
@@ -782,102 +783,187 @@ void ACesium3DTileset::MarkTilesetDirty() {
...
+    pCameraManager->GetCameraViewPoint(location, rotation);
+    double fov = pCameraManager->GetFOVAngle();
+
+    FVector2D size;
+    pViewport->GetViewportSize(size);
+
+    if (size.X < 1.0 || size.Y < 1.0) {
+      continue;
+    }
+
```
Comment (javagl) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:845]: There are several `if (....) continue;` blocks above. 
Are these anticipated cases? 
If so, a comment like `// This happens when (PIE, loading a level, switching views, ... whatever)` could be helpful. 
If not, may they be worth a warning like `LOG(Warning, "Whoopsie, why is there no CameraManager"`)`?

**Comment by javagl** [comment]
I dropped some small nitpicks inline, and will have a closer look at the `cesium-native` side ASAP.

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp (hunk 1 above)
Comment (nithinp7) [reply Source/CesiumRuntime/Private/Cesium3DTileset.cpp:942]: I initially used GetViewRotation, but it seemed to return FRotator(0, 0, 0). At least when testing with orthographic editor viewports.

**Comment by CesiumBen** [comment]
> @nithinp7 Awesome work and demo!
> 
> I've tested these updates with VR and motion and am still getting some tiles unloading.....

moved to [issue](https://github.com/CesiumGS/cesium-unreal/issues/559)

**Comment by javagl** [comment]
Does this have to be updated for https://github.com/CesiumGS/cesium-native/commit/cc9656230ac26892ac700232bdee87937fa9aca3 ?

**Comment by javagl** [comment]
I tried this out with the latest state and the matching `cesium-native` state. Of course, it is hard to "verify" any form of "correctness" in terms of the selected tiles. But the "Dynamic Pawn" preview window shows the full tileset, even when the main window shows a part that the pawn does *not* look at, which is the most apparent difference here. The things that I tried out also worked well, including the creation of multiple pawns and selecting them (showing multiple preview windows accordingly).

**Comment by nithinp7** [comment]
Hey @CesiumBen, thanks for trying this PR out. I did some more digging and found that Unreal doesn't seem to use two separate cameras for VR as I had hoped, instead there is dedicated stereographic rendering logic. In the last commit I just pushed, I attempted to use the stereographic information to determine the appropriate viewports for the two eyes in VR. When you get time, please let me know if it works, since I currently cannot test it out with VR.

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 4]
```diff
@@ -782,102 +785,164 @@ void ACesium3DTileset::MarkTilesetDirty() {
...
+        static_cast<FEditorViewportClient*>(pViewportClient);
+
+    if (!pEditorViewportClient) {
+      continue;
+    }
+
+    const FVector& location = pEditorViewportClient->GetViewLocation();
+    const FRotator& rotation = pEditorViewportClient->GetViewRotation();
+    double fov = pEditorViewportClient->FOVAngle;
+    FIntPoint _;
```
Comment (baothientran) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:934]: I think we should spell out variable name even though it's unused

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 5]
```diff
@@ -782,102 +785,164 @@ void ACesium3DTileset::MarkTilesetDirty() {
...
+
+  TSharedPtr<IStereoRendering, ESPMode::ThreadSafe> pStereoRendering = nullptr;
+  if (GEngine) {
+    pStereoRendering = GEngine->StereoRenderingDevice;
   }
 
-  return UnrealCameraParameters{size, location, rotation, fov};
+  std::vector<ACesium3DTileset::UnrealCameraParameters> cameras;
+
+  for (auto controllerIt = pWorld->GetControllerIterator(); controllerIt;
```
Comment (baothientran) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:834]: it seems like pWorld has GetPlayerControllerIterator() too. Should we use that to remove the cast below? I'm curious what is the difference between GetPlayerControllerIterator() and GetControllerIterator()?

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 6]
```diff
@@ -782,102 +785,164 @@ void ACesium3DTileset::MarkTilesetDirty() {
...
+    return {};
+  }
+
+  TSharedPtr<IStereoRendering, ESPMode::ThreadSafe> pStereoRendering = nullptr;
+  if (GEngine) {
+    pStereoRendering = GEngine->StereoRenderingDevice;
   }
 
-  return UnrealCameraParameters{size, location, rotation, fov};
+  std::vector<ACesium3DTileset::UnrealCameraParameters> cameras;
```
Comment (baothientran) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:832]: pWorld has GetNumControllers() and GetNumPlayerControllers(), so they can use to reserve the size of cameras vector to minimize allocation

File: Source/CesiumRuntime/Private/Cesium3DTileset.cpp [hunk 7]
```diff
@@ -1158,22 +1223,23 @@ void ACesium3DTileset::Tick(float DeltaTime) {
...
-  Cesium3DTiles::ViewState tilesetViewState =
-      this->CreateViewStateFromViewParameters(
-          camera.value().viewportSize,
-          camera.value().location,
-          camera.value().rotation,
-          camera.value().fieldOfViewDegrees);
+  glm::dmat4 unrealWorldToTileset = glm::affineInverse(
+      this->GetCesiumTilesetToUnrealRelativeWorldTransform());
+
+  std::vector<Cesium3DTiles::ViewState> frustums;
```
Comment (baothientran) [inline Source/CesiumRuntime/Private/Cesium3DTileset.cpp:1234]: frustum's vector should be reserved since we already know the number of camera

**Review by baothientran — COMMENTED** [review]
@nithinp7 this is looking good to me! For the VR camera, unless it can be tested and verified soon, can you separate that commit into a different PR?

**Comment by CesiumBen** [comment]
@nithinp7 Tested the latest commit with no success Testing notes updated on [issue](https://github.com/CesiumGS/cesium-unreal/issues/559#issuecomment-886953723)

**Comment by nithinp7** [comment]
@CesiumBen I was able to reproduce some of the weird culling issues you noticed by using the `-emulatestereo` commandline argument. I think the last commit seems to grab the correct fovs for stereoscopic rendering, so at least the introduced culling bug should be fixed (let me know if it hasn't been). Also please let me know if the issue in god-mode where the frustum only covers one eye is gone, hopefully all the tiles visible in both eyes are now being loaded. If it seems to work for XR, we will likely be trying to get this into the August 1st release.

**Pull request merged**